### PR TITLE
Chat team MVP4 part 1: block user

### LINKED ISSUE #7157: MVP 4 Block users
# Problem

An overview of the background required to understand the problem.
A problem description.

# Implementation

Known steps towards feature implementation.
What needs further specifying and investigating.

# Acceptance Criteria

Rules for the future PR to be accepted.

# Notes

Random notes to keep in mind while implementing it. Mostly about related issues and future plans and thoughts.

# Future Steps

Steps which should be taken after this issue has been resolved.

### PR DESCRIPTION
**Pull request by yenda**
partially fixes #7157

### Summary:
block

- user can block other users from their profile page
- all messages from blocked user are deleted and future ones are ignored
- chats are updated accordingly, updating last-message and unread counter that are denormalized and recomputing message groups of affected chats

Currently the user can only been unblocked before you leave the profile. Later we will reintroduce contact list with a list of blocked users so that users can be unblocked

#### Platforms
- Android
- iOS

#### Areas that maybe impacted (optional)
**Functional**
- 1-1 chats
- public chats
- group chats

status: ready

### CODE REVIEW

File: src/status_im/contact/core.cljs
Comment (rasom) [inline src/status_im/contact/core.cljs]: `chat-id` is never used

File: src/status_im/contact/core.cljs
Comment (rasom) [inline src/status_im/contact/core.cljs]: `message-id` is never used

File: src/status_im/contact/core.cljs
Comment (rasom) [inline src/status_im/contact/core.cljs]: `db` is never used

File: src/status_im/contact/core.cljs
Comment (rasom) [inline src/status_im/contact/core.cljs]: `chat` is never used

File: src/status_im/data_store/contacts.cljs
Comment (rasom) [inline src/status_im/data_store/contacts.cljs]: We never know how many messages there, and `core/all-clj` might be quite expensive if thousands. As far as I understand we only need `message-id` and `chat-id`, so we can directly get them from js object, for instance via [`.reduce`](https://realm.io/docs/javascript/latest/api/Realm.Collection.html#reduce).

File: src/status_im/data_store/contacts.cljs
Comment (rasom) [inline src/status_im/data_store/contacts.cljs]: as long as `:chat` is not used later probably no need to execute this query here

File: src/status_im/data_store/contacts.cljs
Comment (rasom) [inline src/status_im/data_store/contacts.cljs]: we might want to reuse the collection from `get-blocked-user-data` here, depending on the number of messages for this specific  `public-key` and the total number of messages in db this query might be expensive, worth checking

File: src/status_im/data_store/contacts.cljs
Comment (rasom) [inline src/status_im/data_store/contacts.cljs]: We can also try to fetch statuses by primary key because we do know them here, it might be faster. Depends on quantities

File: src/status_im/contact/db.cljs
Comment (rasom) [inline src/status_im/contact/db.cljs]: can be just `(filter :blocked? contacts)`

**Comment by rasom** [comment]
We can also simulate some kind of spam attack and check how these changes are useful in that case.

File: src/status_im/data_store/contacts.cljs
Comment (rasom) [inline src/status_im/data_store/contacts.cljs]:
```
ReactNativeJS: ERROR [status-im.utils.async:19] - failed to run task Error: Object type 'messages' not found in schema.
```

File: src/status_im/data_store/contacts.cljs
Comment (rasom) [inline src/status_im/data_store/contacts.cljs]: this is query is going to be slower than fetching each separate entry using [`.objectForPrimaryKey`](https://realm.io/docs/javascript/latest/api/Realm.html#objectForPrimaryKey)

Comment (yenda) [reply src/status_im/data_store/contacts.cljs]: we'll leave that kind of optimization for later this is a mvp and at that point the visible effects of blocking are already applied so perf doesn't matter that much.

**Comment by rasom** [comment]
In general, what's the goal of this particular PR: to give the user an ability to mute an annoying user or to protect from some kind of dos attacks?

File: src/status_im/data_store/contacts.cljs
Comment (rasom) [reply src/status_im/data_store/contacts.cljs]: i agree, but also i would at least consider adding activity indicator because it took 10s on my slower device to block the user which already has sent 1000 messages.

Comment (rasom) [reply src/status_im/data_store/contacts.cljs]: 6s for removing 6 messages from 7 chats with ~2000 messages in total.

**Comment by yenda** [comment]
@rasom for now it's to block the kind of spam we had a couple of time already with someone posting useless huge messages.

File: src/status_im/data_store/contacts.cljs
Comment (yenda) [reply src/status_im/data_store/contacts.cljs]: What takes 10s is it the realm part or the whole thing. I would think recomputing the message groups of the affected chats is more expensive. Also this is the MVP of the MVP so I'm not really concerned about slow devices at that point

**Comment by rasom** [comment]
@yenda with the assumption that it will be hard for this person to automate account creation and to send N messages per acc? It's kind of easy thing to implement and in this case, this functionality will be useless. If we just want to introduce user blocking functionality then fine, but we if need spam protection this thing isn't a big obstacle.

**Comment by yenda** [comment]
Yes it's just to block a user

File: src/status_im/data_store/contacts.cljs
Comment (rasom) [reply src/status_im/data_store/contacts.cljs]: most likely realm, because otherwise, each incoming message would have a similar effect. I doubt that those subs have any serious impact here.

It is 1.2-1.5s on galaxy s9 which is not a slow device, that's when removing messages 6 messages with 6 chats and ~2000 messages in total. A realistic test would require 20-30 chats, 15-20k messages in the realm db and removing dozens of messages from different chats. I suppose it might be up to 2.5-3s. 

Well if that's not necessary from UX pov then no problem.

**Review by rasom — APPROVED** [review]

**Review by vitvly — COMMENTED** [review]
@yenda I've noticed that there is no desktop support yet - is this for future versions of this feature? (desktop's profile screen code hasn't been changed https://github.com/status-im/status-react/blob/b56fd2e29fae97c47a8274449adaa70526665e3b/src/status_im/ui/screens/desktop/main/chat/views.cljs#L348)

**Comment by yenda** [comment]
@siphiuel yes it's for a future iteration

**Review by vitvly — APPROVED** [review]

**Comment by rachelhamlin** [comment]
Check: is this reversible if I know a user's contact ID/username, and look up their profile?

**Comment by yenda** [comment]
@churik check what happens when trying to add again a blocked user using his address/ens username

**Review by errorists — APPROVED** [review]

**Comment by churik** [comment]
@yenda 
**1) After reopening app you again will receive messages from blocked contact**

Steps:
- log in (user A)
- join public channel (user A)
- send any message to public channel (user B)
- block user B (user A)
- reopen app (user A)
- send message again to public channel (user B)

**Expected result:** User A can't see messages from User B, because he is blocked
**Actual result:** User A can see messages from User B
Log:  [logcat.txt](https://github.com/status-im/status-react/files/2744718/logcat.txt)
If you log out when switching networks, all blocked messages are reappering.

**2) Block contact from 1-1 chat header leads to 'Null is not an object' error**
Steps:
- log in (user A)
- join public channel (user A)
- send any message to public channel (user B)
- open User B profile (user A)
- start chat with user B (user A)
- in 1-1 chat with user B tap on three dots in header > view profile > Block user (user A)
- go back (user A)

**Expected result:** no error, user A is redirected to chat list
**Actual result:**  
nullisnotanobj
Same error at attempt to add to contacts blocked user.


**3) After blocking and unblocking user I don't see messages from user**
Steps:
- log in (user A)
- join public channel (user A)
- send any message to public channel (user B)
- open User B profile (user A)
- block user B (user A)
- unblock user B (user A)
- go back to public chat

**Expected result:** can see messages from User B
**Actual result:** can't see messages from User B

If unblock feature is not working yet, we need to disable it IMO.

**4) Q: if I'll have the same account on several devices, should I block user on all of them separately? I suppose that it should be  automatically synced, but it will be separate task, am I right?**

**Comment by churik** [comment]
@yenda ping me please when PR will be ready to test

**Comment by yenda** [comment]
@churik it is ready to test

**Comment by churik** [comment]
@yenda 
1) is still reproducible, can receive notification from blocked contact after reopening app (tried on IOS [build](https://i.diawi.com/2pcaxW) )
2) still reproducible after app relaunch (tried on IOS [build](https://i.diawi.com/2pcaxW) )

**Comment by yenda** [comment]
@churik ok it should be fixed somehow the change I made to fix it disappeared which is why I thought it was working

**Comment by yenda** [comment]
@churik regarding notifications please create an issue and mention me and @PombeirP we don't need to fix it now but we should keep it in mind.

**Comment by churik** [comment]
@yenda 
1 and 2 are still reproducible if chat with blocked user contains messages from both sides (blocked user and blocking user).

Modified steps:
- log in (user A)
- join public channel (user A)
- send any message to public channel (user B)
- open User B profile (user A)
- start chat with user B (user A)
- send several messages to 1-1 chat (user A)
-  send several messages to 1-1 chat (user B)
- in 1-1 chat with user B tap on three dots in header > view profile > Block user (user A)
- go back (user A)

1) Null is not an object error is still there
2) After relaunching app User A can see messages from User B again in all chats

**Comment by churik** [comment]
@rachelhamlin 
Want to confirm that all of these items are OK for part 1: 
- you can't add user to contact after blocking, at attempt there is no validation message, nothing happens;
- after blocking user you still receive PNs if he writes you a message but you don't see a message itself;
- no desktop support - so you can't block user from desktop;
- blocking action is irreversible;
- there is no sync for blocked contact, so in case if you are using several devices and one of them is desktop you can't bock contact and still will receive messages from blocked contact in desktop.

**Comment by rachelhamlin** [comment]
Let's discuss on call tomorrow, but I'd like to have a plan for porting this feature to desktop & also understand the effort to make it reversible (with a blocked users list).

About the notifications. Phantom notifications were once a bug that we specifically put energy towards solving, so I don't love introducing a new form of it with this feature.

That said, the feature is most useful for blocking annoying people in public chats. I am rarely spammed 1:1, and TtT will soon be in place to help with that, preventing phantom notifications coming in from users I've blocked.

**Comment by pedropombeiro** [comment]
> That said, the feature is most useful for blocking annoying people in public chats. I am rarely spammed 1:1, and TtT will soon be in place to help with that, preventing phantom notifications coming in from users I've blocked.

FWIW, for the [PN metadata leak issue](https://github.com/status-im/status-react/issues/6772), I recently added support for ignoring PNs if they come from users which are not on your contact list.

**Comment by rachelhamlin** [comment]
@PombeirP great to know, thanks. That should fix that issue then.

**Comment by yenda** [comment]
@churik have you noticed the latest update?

**Comment by churik** [comment]
@yenda
**2) Block contact from 1-1 chat header leads to 'Null is not an object' error** is still reproducible, steps are changed a bit. Now issue is reproducible if both users are added as contacts of each other.
Updated steps:

- log in (user A)
- add user B as contact (user A), send several messages to 1-1 chat
- add User A as contact (user B),  send several messages to 1-1 chat
- open User B profile (user A)
- start chat with user B (user A)
- in 1-1 chat with user B tap on three dots in header > view profile > Block user (user A)
- go back (user A)
img_a9a09dac09ef-1

**Comment by churik** [comment]
@yenda 
### 1) "Contacts" in Profile doesn't count blocked contacts
bl_contacts

### 2) On "Contacts" screen:
2.1. There is enough space to show "Blocked contacts" in full instead of "Blocked conta..."
2.2. "Add to contacts" doesn't work (no action on tapping)
bl_cjd

### 3) Icons of assets are changed
exp


### 4) At attempt to open Group info receive Error: Unknown icon

Steps:
- log in
- open group chat
- Tap on 3 dots>group info 
error

### 5) No new messages in group chat from unblocked user
Steps:
- as user A: create group chat, add user B to group chat
- as user A: block user B 
- as user A: unblock user B
- as user B: send message to group chat

_Expected result:_ as user A: can see new message from user B
_Actual  result:_ no new messages

### 6) Also from my side it is not clear why blocked user can't receive messages in group chat - is it an issue?

**Comment by yenda** [comment]
1) that's normal, a blocked contact doesn't count as a contact, some aren't even contacts before being blocked
2) looking into it
3) can't reproduce that, is it on iOS?

**Comment by churik** [comment]
1. they are placed inside Contacts, so seems strange when you open Contacts(0) and inside see Blocked contacts(2). WDYT?
3. yes, IOS only

**Comment by churik** [comment]
@adan - can you please align tests with changes in this branch? 'ShareMyProfileButton'  element is changed.
Thanks!

**Comment by yenda** [comment]
1) "Contacts" in Profile doesn't count blocked contacts
In the design there isn't an example that answers that question but personally I wouldn't want the blocked contacts to be included in the contacts count.

2) Should be fixed now but I can't test because I don't have a phone with a screen width small enough to trigger this.

3) Icons of assets are changed
Should be fixed

4) At attempt to open Group info receive
Fixed

**Comment by antdanchenko** [comment]
accessibility label `share-my-profile-button` has been removed for `Share my profile` button with redesign
@yenda  Please, restore the label for the button

**Comment by churik** [comment]
@yenda as we discussed on Fraiday call, "Blocked contacts" should be renamed to "Blocked users" to avoid confusion. Please ping me when I can retest build with renamed label and fixed https://github.com/status-im/status-react/pull/7170#issuecomment-457765828.
Thanks!

**Comment by yenda** [comment]
@churik it's changed already

**Comment by yenda** [comment]
@churik also I made a small change that might be sufficient to have blocked? contact sync between device could you check if it works?

**Comment by churik** [comment]
@yenda 
**1) When accessing  blocked users list, translation is missing in header**
Screen:
missing


Testing in progress

**Comment by yenda** [comment]
@churik I fixed the label, the accessibility label and rebased on develop

**Comment by churik** [comment]
@yenda 
### 1) Sync of blocked users doesn't work if you have 1-1 chat with him in last 24 h

**Steps:**
- device A: create new account
- device A: add several contacts ("UserA" and "UserB")
- device A: with one of contacts (let's say, contact "UserA") start chat. As UserA, send several messages to chat
- device A: block both "UserA" and "UserB"
- device B: restore same account from seed phrase. Chat with  "UserA" will be restored.
- device B: go to Profile > Development mode > Devices > Set name for device > Tap on +
- device A: go to Profile > Development mode > Devices > Set name for device > Tap on +
- device A: enable sync with device B
- device B: check Contact list and blocked users list

**Expected:** "UserA" and "UserB"  are blocked as on device B
**Actual:** "User A" is restored as contact

Log from device A: [geth.log](https://github.com/status-im/status-react/files/2807261/geth.log)
Log from device B: [logcat.txt](https://github.com/status-im/status-react/files/2807262/logcat.txt)


Also sync is not happening automatically and don't work when you unblock users.
So basically sync is working only when you manually tap on "sync devices" and blocked user wasn't added to your contact list - in this case he will be added to "Blocked users" on another device.


### 2) Icon error when you tap on stickers
icon

**Comment by churik** [comment]
@yenda also please resolve conflicts

@antdanchenko I think that  `test_backup_recovery_phrase` and `test_recover_account` can fail because button "Backup recovery phrase" now is placed below visible area.
Not sure about first failure though. Can you please clarify?

**Comment by yenda** [comment]
@churik so I propose that we remove the sync from this PR so that we can merge it once release has been cut. I'll work on it separately

**Comment by churik** [comment]
@rachelhamlin 
Please, confirm that it is OK if we merge to develop this feature (after release branch will be cut off) with such known issues/behavior:
- no sync of "Blocked users" across devices
- no desktop support
- If some **user A**  blocks by **User B** -> then **User B** creates a chat and invite **User A** and **User C** -> **User A** will receive an invite after **User C** will send the first message to a chat
- if you manually try to add blocked user as a contact after you had chat with him -> you will be redirected to chat with own messages only
- if you block user from "Blocked contacts" screen -> instead of profile you'll be redirected to Home screen
-  you'll receive PNs from blocked user

@yenda at least we should have 100% Passed e2e tests in this PR

**Comment by antdanchenko** [comment]
@churik please do not merge until getting a commit with fixes for tests, ~30 minutes

**Comment by rachelhamlin** [comment]
@churik thanks - I think we can live with those issues in the MVP. The biggest one is syncing/desktop imo. I'm going to add stories to our pivotal icebox for desktop support + bugs so we can prioritize later.

**Comment by yenda** [comment]
@rachelhamlin please document in the chat notes first, we can create this issues on pivotal during Friday call. We already started with @churik https://notes.status.im/weekly-chat-sync-agenda?edit

The plan is that we have everything documented, then create the issues and commit current behavior in the repo

<blockquote><div><strong><a href="https://notes.status.im/weekly-chat-sync-agenda?edit">CodiMD - Collaborative markdown notes</a></strong></div><div>The best platform to write and share markdown.</div></blockquote>

**Comment by antdanchenko** [comment]
@yenda I've added a commit which will fix 2 tests, - [scroll to BackupRecoveryPhraseButton](https://github.com/status-im/status-react/pull/7170/commits/63c90a649aced13978daff114cd25b4b2e43b022), please mind the commit before squash

**Comment by churik** [comment]
@yenda please, rebase it one more time - got fix for e2e https://github.com/status-im/status-react/pull/7381

**Comment by churik** [comment]
@yenda 

**1) 'Null is not an object' error is still reproducible**

Updated steps:
- log in (user A)
- add user B as contact (user A), send several messages to 1-1 chat
- add User A as contact (user B),  send several messages to 1-1 chat
- open User B profile (user A)
- start chat with user B (user A)
- in 1-1 chat with user B tap on three dots in header > view profile > Block user (user A)
- close app (user A)
- send several messages to user A (user B)
- open app (user A)

After opening app User A redirected to chat with blocked user
img_a9a09dac09ef-1
Same is happening if you tap on PN from blocked user.


**2) After blocking contact messages from blocked user are preserved in 1-1 chat after adding him as a contact**

On previous builds all messages sent before blocking were deleted -  so I suppose this is the right behavior.
For now if you add user after blocking him - all previous messages are preserved and chat is created again
Video: http://take.ms/CICVi
On "Home" screen "No messages" is shown, but if chat is opened - all history is preserved.
dsf

**3) Messages from blocked contact in public chat are restored after reopening app**
Steps:
- open any public chat 
- open any user profile >  block user
- close app
- open app and check public chat
**Expected result:** no messages from blocked user
**Actual result:** all messages from blocked user are restored in public chat

Sync is disabled and all blocked users are automatically added as contacts after sync.

**Comment by churik** [comment]
@antdanchenko check please, `test_add_contact_from_public_chat` failure should be fixed by you or by Eric.
Thanks!

**Comment by yenda** [comment]
@antdanchenko please don't use force push you are overwriting my change every time

**Comment by antdanchenko** [comment]
@yenda I did git rebase origin/mvp4-1-block-user and then just simple push, no force push 
screen shot 2019-01-31 at 12 46 16 pm

**Comment by yenda** [comment]
@antdanchenko ok my bad, it's because I removed a commit but you added it back twice with your push

**Comment by yenda** [comment]
@churik all fixed

1) was caused by the notification trying to navigate to a deleted chat. the fix was to verify that the contact isn't blocked before navigating to the chat mentioned in a notification

2) was caused by a wrong conditional I introduced while refactoring the code for cleaning up the db when blocking users to make it more feasible and fix the previous version of the bug (`empty` instead of `empty?` which was always `true` so never deleted the messages)

3) same

**Comment by churik** [comment]
@yenda after creating two users on android and IOS and continue messaging after I sent to chat sticker I've got error:
img_2567b11f3918-1
So steps to reproduce:
- start 1-1 chat with user B
- as User B add sticker and send one to 1-1 chat
- send one more message 
- check User A device
Not reproducible on nightly or other PR builds.

Also want to retest this PR when https://github.com/status-im/status-react/issues/7410 will be fixed and included on develop

**Comment by yenda** [comment]
@churik should be fixed

**Comment by churik** [comment]
@yenda please rebase to include https://github.com/status-im/status-react/pull/7412

**Comment by yenda** [comment]
already done

**Comment by churik** [comment]
@yenda @rachelhamlin 
**1) if user tap on PN from blocked user it will be redirected to "Unknown" chat,** from where you can navigate to Unknown user profile and even add Unknown to Contacts. And if you continue work with this Unknown contact, for example block him -> eventually you will see TypeError: null is not an object error




**2) In public chat: after blocking user and hiding his messages,  "No messages" is shown on "Home" screen**
home3

**3) Can add blocked users to a group chat**
But when blocked user will send a message, you will not receive it.
Will add this to known issues for part1 ?

**Comment by rachelhamlin** [comment]
@churik yes please. I imagine this function is most useful for removing users from public chats at this point, so it's okay to fix 1:1 issues in next iteration.

**Comment by churik** [comment]
Tested:
* blocking user which have messages in 1-1, public, group chat -> send messages to all of them and reopening app - no messages are appeared;
* blocking user from public, 1-1, group info and Contacts screen;
* send message from blocked contact from different devices;
* checking that impossible to receive invite to group chat from blocked user -> but user will receive invite when other user will join public chat;
* changing networks/fleets;
* checked blocking user and hiding all messages from him if chat has > 24h history
* blocking user from release 0.9.32;
* restore account with blocked users and pairing between several devices (no sync for blocked users now) and blocking the same user on each device;
* add user while it is blocked;
* PNs from blocked users (known issue);
* Blocking-unblocking users several times;
* `/send` and `/request` commands 
* upgrade from 0.9.32 with  group chat, 1-1, several users in contact list, "Add to contact" in 1-1 - can use blocking feature after upgrade

**Comment by churik** [comment]
So known behavior for now:
- no sync of "Blocked users" across devices
- no desktop support
- If some **user A**  blocks by **User B** -> then **User B** creates a chat and invite **User A** and **User C** -> **User A** will receive an invite after **User C** will send the first message to a chat
- if you manually try to add blocked user as a contact after you had chat with him -> you will be redirected to chat with own messages only
- if you block user from "Blocked contacts" screen -> instead of profile you'll be redirected to Home screen
-  you'll receive PNs from blocked user and if you open it - "Unknown" chat is shown
- In public chat: after blocking user and hiding his messages,  "No messages" is shown on "Home" screen
home3
- can add blocked users to a group chat. But when blocked user will send a message, you will not receive it.

**Comment by yenda** [comment]
@churik so we can merge this?
I will prioritize the PN problem, I tried to reproduce the "no messages" in public chat but couldn't, is there further steps to take?

**Comment by churik** [comment]
@yenda yes, you can merge it!
I can create existing issues to keep track on them with separate label, i.e. "block", would it be helpful?

**Comment by yenda** [comment]
@churik yeah sure but please create them in pivotal as well since we are traking progress there now

**Pull request merged**
